### ISSUE TITLE
feature suggestion

### ISSUE THREAD

**Issue by Fuseteam** [salty-horse/ibus-uniemoji#25]
Having the emoji picker appearing all the time while typing might be cumbersome at times
It would be useful if it only appeared when the user needs emoji
A common way i've seen this implemented is by using ":" not unlike how it works on github :smile: 
![image](https://user-images.githubusercontent.com/10421851/37948224-8f0dc45a-3165-11e8-850c-79ece4a7de89.png)

**Comment by Depau** [comment]
You can switch to another input method when you don't want the picker to appear. On GNOME you can add multiple input methods and use `Super`+`Space` to switch among them.

**Comment by salty-horse** [comment]
I agree with Depau.

I'm not sure what you want is possible with ibus. It sounds you wish to keep the current language's input method, but automatically switch to emoji-input whenever `:` is pressed.

**Comment by Fuseteam** [comment]
What I'm suggesting is that the suggestion box is triggered using ":"
As switching IME can be hassle if one already used multiple IME's

On Mar 27, 2018 6:18 AM, Ori Avtalion <notifications@github.com> wrote:

I agree with Depau.

I'm not sure what you want is possible with ibus. It sounds you wish to keep the current language's input method, but automatically switch to emoji-input whenever : is pressed.

**Comment by salty-horse** [comment]
Are you referring to a situation where you want to type an emoji, then type text, then type another emoji?

What should happen when you type ESC after entering this mode? Should it leave the text you used to search an emoji as "real text", or should it delete it? I can see how both cases can be useful and/or annoying in different situations.

**Comment by Depau** [comment]
I think what he's trying to propose is that normally the IME should insert the text directly into the field without showing the dialog.
I.e. typing `joy` won't type the `😂` emoji, would just insert "joy". In order to type it one should type `:joy` and then the menu will appear (and maybe not show up when typing `: ` — semicolon then space).

I think it could be a useful feature, if made optional, as most of us users are already used to `Super`+`Space`, then `text`, `Return`, `Super`+`Space` when typing an emoji.

**Comment by Fuseteam** [comment]
yes i'm trying to propose what depau said basically a type of prefix search function
its very useful
its also pretty much how emoji input works here on github as you see in the picture but also how it works in apps like telegram and discord :joy:

**Comment by salty-horse** [comment]
I don't think @Depau accurately describes what you want since Github et al still show the drop-down.

I have several questions now:
1. You do understand you still need to switch to the emoji input method...? So entering the : will just slow things down (unless you want to still be able to combine it with typing text that won't be caught by the search)
2. In all websites, there's no standard to what Unicode characters keywords after ":" are mapped to. UniEmoji uses EmojiOne's mapping of such "shortnames". After entering a `:`, do you wish to be able to search ALL UniEmoji entries, and not just shortnames?

**Comment by salty-horse** [comment]
Also, how can you search multiple keywords in `:` mode? In Github, for example, pressing a space "cancels" the dropdown but keeps the text you wrote.

**Comment by Fuseteam** [comment]
yes i'm not saying we don't need the drop down 
basically i'm proposing instead of triggering this search drop down by simply typing 
it could be triggered by typing ":" first
you can search multiple keywords by using another symbol in github's case tis "_"

**Comment by salty-horse** [comment]
underscores are already part of shortnames on Github. Github it's not really a search and it's not really a keyword separator.

If you're suggesting ignoring underscores in names, and instead using them as a keyword separator, that's possible.

**Comment by Fuseteam** [comment]
would space be a problem if the search is triggered by a colon tho? or an other character/key cause that's what i'm basically suggesting: a way to activate or deactivate emoji input without switching keyboards
i'm suggesting ":" because i'm basically already used to using that on mobile and on some sites

**Comment by salty-horse** [comment]
It's possible to have space remain in search mode, which is what the current behavior does. I was suggesting it can change to better emulate Github's behavior, but I think it's better to keep it as-is.

OK, I will try implementing the `:` to trigger the search. Also need to create a config file to toggle it.

**Comment by Fuseteam** [comment]
in the meantime i'll remove the `:` from my pull request since it doesn't appear to be needed there

**Comment by Fuseteam** [comment]
i have been trying hack an trigger, so far without any luck...

**Comment by fpigeonjr** [comment]
> Having the emoji picker appearing all the time while typing might be cumbersome at times
> It would be useful if it only appeared when the user needs emoji
> A common way i've seen this implemented is by using ":" not unlike how it works on github 
> ![image](https://user-images.githubusercontent.com/10421851/37948224-8f0dc45a-3165-11e8-850c-79ece4a7de89.png)

this would be a killer feature. I am looking for something like the [rocket on mac](https://matthewpalmer.net/rocket/)

**Comment by Fuseteam** [comment]
this would still be a killer feature imo
for now i have found one clunky work around with a script
```
#!/bin/bash
  1 engine=`ibus engine`
  2 if [ "$engine" = "uniemoji" ]; then
  3         ibus engine $oengine
  4 else
  5         export oengine="$engine";ibus engine uniemoji
  6 fi
```
toggling this script sets the input method to uniemoji and back to whatever the user had first
the issue is the terminal where this script is ran cannot be closed :x

**Comment by fuzxi** [comment]
@Fuseteam 
> the issue is the terminal where this script is ran cannot be closed :x

Why don't you just bind the script to a key, using your WM or `bind`?

**Comment by Fuseteam** [comment]
i did, it switches to uniemoji but it doesn't switch back.
haven't found a work around for that
would still be better if we can integrate this functionality i.e. toggling between typing emoji and typing normally

**Comment by Fuseteam** [comment]
i just had the idea to perhaps cancel emoji input and just leave the inputted text using the tab key, since that key seems unused atm ☺️

**Comment by rishikesh2003** [comment]
is this feature enabled yet?

**Comment by salty-horse** [comment]
No, I didn't implement it. Patches are welcome.

**Comment by bbb651** [comment]
> I think it could be a useful feature, if made optional, as most of us users are already used to `Super`+`Space`, then `text`, `Return`, `Super`+`Space` when typing an emoji.

This really doesn't play nice with multiple with any input sources but just English. Switching from `English -> Uniemoji` it works as expected, but from `Hebrew -> Uniemoji` it leaves you in a weird state where `Uniemoji` is the selected input source, but it doesn't actually do anything and the real input source that is used is `Hebrew`, so you have to switch to `"Uniemoji" -> English -> Uniemoji`. I'm not sure if it's an issue in GNOME, ibus or ibus-uniemoji (maybe this can support multiple languages for searching emojis, I personally won't find it useful and unfortunately people who do probably won't ever get to use this as it's not very accessible to find and install for non English speakers...) but it really hurts usability. 

> i just had the idea to perhaps cancel emoji input and just leave the inputted text using the tab key, since that key seems unused atm ☺️

For me this seems very unintuitive as <kbd>Tab</kbd> is usually used as "Complete" in these sort of emoji pickers (e.g. Discord) and code editors. I personally don't see a need to cancel the emoji prompt and leave the inputted text **unless** you just typed <kbd>:</kbd> (<kbd>Space</kbd> OR <kbd>Enter</kbd>) which will often happen when using a colon normally.
